subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/FileSystemChangeListener.java

```diff
@@ -0,0 +1,281 @@
+            pendingChangesListener.onPendingChanges();
+        }
+    }
+
+    @Override
+    public synchronized void onExecute(TaskInternal task, FileCollectionInternal fileSystemInputs) {
+        Set<String> taskInputs = new LinkedHashSet<>();
+        Set<FilteredTree> filteredFileTreeTaskInputs = new LinkedHashSet<>();
+        fileSystemInputs.visitStructure(new FileCollectionStructureVisitor() {
+            @Override
+            public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
+                contents.forEach(location -> taskInputs.add(location.getAbsolutePath()));
+            }
+
+            @Override
+            public void visitGenericFileTree(FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                fileTree.forEach(location -> taskInputs.add(location.getAbsolutePath()));
+            }
+
+            @Override
+            public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
+                if (patterns.isEmpty()) {
+                    taskInputs.add(root.getAbsolutePath());
+                } else {
+                    filteredFileTreeTaskInputs.add(new FilteredTree(root.getAbsolutePath(), patterns));
+                }
+            }
+
+            @Override
+            public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                taskInputs.add(file.getAbsolutePath());
+            }
+        });
+        inputAccessHierarchy.recordInputs(taskInputs);
+        filteredFileTreeTaskInputs.forEach(fileTree -> inputAccessHierarchy.recordFilteredInput(fileTree.getRoot(), fileTree.getPatterns().getAsSpec()));
+    }
+
+    private static long monotonicClockMillis() {
+        return System.nanoTime() / 1000000L;
+    }
+
+    private static class FileEventCollector {
+        private static final int SHOW_INDIVIDUAL_CHANGES_LIMIT = 3;
+        private final Map<Path, FileWatcherRegistry.Type> aggregatedEvents = new LinkedHashMap<>();
+        private int moreChangesCount;
+        private boolean errorWhenWatching;
+
+
+        public void onChange(FileWatcherRegistry.Type type, Path path) {
+            FileWatcherRegistry.Type existingEvent = aggregatedEvents.get(path);
+            if (existingEvent == type ||
+                (existingEvent == FileWatcherRegistry.Type.CREATED && type == FileWatcherRegistry.Type.MODIFIED)) {
+                return;
+            }
+
+            if (existingEvent != null || aggregatedEvents.size() < SHOW_INDIVIDUAL_CHANGES_LIMIT) {
+                aggregatedEvents.put(path, type);
+            } else if (shouldIncreaseChangesCount(type, path)) {
+                moreChangesCount++;
+            }
+        }
+
+        private boolean shouldIncreaseChangesCount(FileWatcherRegistry.Type type, Path path) {
+            if (IS_MAC_OSX) {
+                return true; // count every event on OSX
```

Comment:
Would be worth adding "why" is that we count every OSX event?

Reply:
Added.